lib/request/complement.js

```diff
@@ -79,96 +77,135 @@ function findPartialIntersections(requestedPath, optimizedPath, requestTree) {
-                            arrayConcat(oCurrentPath, [innerKey])
-                        );
-
-                        intersections = arrayConcat(intersections, intersectionData[0]);
-                        oComplementPaths = arrayConcat(oComplementPaths, intersectionData[1]);
-                        rComplementPaths = arrayConcat(rComplementPaths, intersectionData[2]);
-                    }
-                    innerKey = iterateKeySet(key, note);
-                }
-
-                // The remainder of the path was handled by the recursive call, terminate the loop
-                break;
-            } else {
-                // For simple keys, we don't need to branch out. Loop over `depth` instead of iterating over a range.
-                currentTree = currentTree[key];
-                oCurrentPath[oCurrentPath.length] = optimizedPath[depth];
-                rCurrentPath[rCurrentPath.length] = requestedPath[depth + depthDiff];
-
-                if (currentTree === undefined) {
-                    // The path was not found in the tree, add this to complements
-                    oComplementPaths[oComplementPaths.length] = arrayConcat(
+function recurse(requestedPath, optimizedPath, currentTree, depth, rCurrentPath, oCurrentPath) {
+    var depthDiff = requestedPath.length - optimizedPath.length;
+    var intersections = [];
+    var rComplementPaths = [];
+    var oComplementPaths = [];
+    var oPathLen = optimizedPath.length;
+
+    // Loop over the optimized path, looking for deduplication opportunities
+    for (; depth < oPathLen; ++depth) {
+        var key = optimizedPath[depth];
+        var keyType = typeof key;
+
+        if (key && keyType === "object") {
+            // If a range key is found, start an inner loop to iterate over all keys in the range, and add
+            // intersections and complements from each iteration separately.
+            //
+            // Range keys branch out this way, providing individual deduping opportunities for each inner key.
+            var note = {};
+            var innerKey = iterateKeySet(key, note);
+
+            while (!note.done) {
+                var nextTree = currentTree[innerKey];
+                if (nextTree === undefined) {
+                    // If no next sub tree exists for an inner key, it's a dead-end and we can add this to
+                    // complement paths
+                    oComplementPaths[oComplementPaths.length] = arrayConcatSlice2(
                         oCurrentPath,
-                        arraySlice(optimizedPath, depth + 1)
+                        innerKey,
+                        optimizedPath, depth + 1
                     );
-                    rComplementPaths[rComplementPaths.length] = arrayConcat(
+                    rComplementPaths[rComplementPaths.length] = arrayConcatSlice2(
                         rCurrentPath,
-                        arraySlice(requestedPath, depth + depthDiff + 1)
```

Comment:
More `concat` of `slice`. I added a utility below to avoid the extra arrays in each of these cases.